code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/deployment/impl/DeploymentManagerImpl.java

```diff
@@ -0,0 +1,364 @@
+package io.cattle.platform.servicediscovery.deployment.impl;
+
+import io.cattle.platform.core.constants.CommonStatesConstants;
+import io.cattle.platform.core.model.Service;
+import io.cattle.platform.engine.idempotent.IdempotentRetryException;
+import io.cattle.platform.lock.LockCallbackNoReturn;
+import io.cattle.platform.lock.LockManager;
+import io.cattle.platform.lock.definition.LockDefinition;
+import io.cattle.platform.object.ObjectManager;
+import io.cattle.platform.object.process.ObjectProcessManager;
+import io.cattle.platform.object.process.StandardProcess;
+import io.cattle.platform.object.util.DataAccessor;
+import io.cattle.platform.servicediscovery.api.constants.ServiceDiscoveryConstants;
+import io.cattle.platform.servicediscovery.api.dao.ServiceExposeMapDao;
+import io.cattle.platform.servicediscovery.deployment.DeploymentManager;
+import io.cattle.platform.servicediscovery.deployment.DeploymentUnitInstance;
+import io.cattle.platform.servicediscovery.deployment.DeploymentUnitInstanceFactory;
+import io.cattle.platform.servicediscovery.service.ServiceDiscoveryService;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+
+import javax.inject.Inject;
+
+import org.apache.commons.lang3.tuple.Pair;
+import jersey.repackaged.com.google.common.collect.Sets;
+
+public class DeploymentManagerImpl implements DeploymentManager {
+    @Inject
+    LockManager lockManager;
+    @Inject
+    ServiceDiscoveryService sdService;
+    @Inject
+    DeploymentUnitInstanceFactory deploymentUnitInstanceFactory;
+    @Inject
+    ObjectProcessManager objectProcessManager;
+    @Inject
+    ObjectManager objectManager;
+    @Inject
+    ServiceExposeMapDao exposeMapDao;
+
+    @Override
+    public void activate(final Service service, final Map<String, Object> data) {
+        final List<Service> services = exposeMapDao.collectSidekickServices(service, null);
+        if (services.isEmpty()) {
```

Comment:
This can never be true, but you should maybe check if service == null, just return as the first thing in this function